File: elixir/scripts/keylog_check.exs
# Reads the shared key-log vector (spec/corpus/keylog.json) and confirms this
# implementation can open a key-logged capture.
#
# security.md §8 pins one implementation-neutral key-log format precisely so a
# single inspector reads a log written by a node in any language. That claim needs
# agreement in BOTH directions: emitting lines your own reader accepts is not
# enough. This checks the reading half against a committed capture the Java
# reference produced. The writing half is covered by each port's own key-log tests
# and, live and cross-language, by interop tier 10.
#
# Invoked by interop/check-keylog-elixir.sh.

alias Bonemesh.Transport

[path] = System.argv()
doc = path |> File.read!() |> JSON.decode!()

capture = doc["capture"] || []
expected = doc["expected"] || []

if capture == [] or length(capture) != length(expected) do
  IO.puts(:stderr, "vector malformed: #{length(capture)} capture, #{length(expected)} expected")
  System.halt(1)
end

# '#' lines are comments; an unknown label shape is ignored rather than fatal, so a
# future label is not a breaking change.
keys =
  (doc["keylog"] || [])
  |> Enum.reduce(%{}, fn raw, acc ->
    line = String.trim(raw)

    cond do
      line == "" or String.starts_with?(line, "#") ->
        acc

      true ->
        case String.split(line) do
          [label, _th, hex] ->
            case Regex.run(~r/^BMX3_(I2R|R2I)_TRAFFIC_(\d+)$/, label) do
              [_, dir, epoch] ->
                case Base.decode16(hex, case: :lower) do
                  {:ok, key} when byte_size(key) == 32 ->
                    Map.put(acc, {String.downcase(dir), String.to_integer(epoch)}, key)

                  _ ->
                    acc
                end

              _ ->
                acc
            end

          _ ->
            acc
        end
    end
  end)

if map_size(keys) == 0 do
  IO.puts(:stderr, "no usable key-log entries in the vector")
  System.halt(1)
end

failures = :counters.new(1, [])

Enum.zip(capture, expected)
|> Enum.with_index()
|> Enum.each(fn {{frame, want}, i} ->
  dir = frame["dir"]
  seq = frame["frame"]["seq"]
  ct = Base.decode64!(frame["frame"]["ct"])

  case Map.get(keys, {dir, want["epoch"]}) do
    nil ->
      IO.puts("FAIL frame #{i}: no key for #{dir} epoch #{want["epoch"]}")
      :counters.add(failures, 1, 1)

    key ->
      case Transport.open_ciphertext(key, seq, ct) do
        {:ok, pt} ->
          # Decoded maps compare structurally, so key order does not matter --
          # the contract is the structure, not the text.
          got = JSON.decode!(pt)

          if got == want["inner"] do
            IO.puts("PASS frame #{i} (#{dir} seq #{seq})")
          else
            IO.puts("FAIL frame #{i}\n  got:  #{inspect(got)}\n  want: #{inspect(want["inner"])}")
            :counters.add(failures, 1, 1)
          end

        :error ->
          IO.puts("FAIL frame #{i}: the logged #{dir} key did not open it")
          :counters.add(failures, 1, 1)
      end
  end
end)

# Self-test the oracle: a ciphertext no key seals must be refused, or a checker
# that reported success for everything would look identical to this one.
{_k, any_key} = Enum.at(keys, 0)

case Transport.open_ciphertext(any_key, 0, <<0::256>>) do
  :error ->
    IO.puts("PASS self-test: an unopenable frame is refused")

  {:ok, _} ->
    IO.puts("FAIL self-test: an unopenable frame was accepted")
    :counters.add(failures, 1, 1)
end

n = :counters.get(failures, 1)

if n > 0 do
  IO.puts(:stderr, "#{n} key-log frame(s) failed")
  System.halt(1)
end

IO.puts("every captured frame opens with its logged key and reproduces the vector")

File: java/src/main/java/com/axonibyte/bonemesh/v3/transport/KeylogCheck.java
/*
 * Copyright (c) 2026 Axonibyte Innovations, LLC. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.axonibyte.bonemesh.v3.transport;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.HashMap;
import java.util.HexFormat;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.json.JSONArray;
import org.json.JSONObject;

/**
 * Reads the shared key-log vector (spec/corpus/keylog.json) and confirms this
 * implementation can open a key-logged capture.
 *
 * <p>security.md §8 pins one implementation-neutral key-log format precisely so
 * that a single inspector reads a log written by a node in any language. That
 * claim needs agreement in <em>both</em> directions: emitting lines your own
 * reader accepts is not enough. This checks the reading half against the
 * committed capture. The writing half is covered by each port's own key-log
 * tests and, live and cross-language, by interop tier 10.</p>
 *
 * <p>Invoked by interop/check-keylog.sh.</p>
 *
 * @author Caleb L. Power
 */
public final class KeylogCheck {

  private KeylogCheck() { }

  private static final HexFormat HEX = HexFormat.of();
  private static final Pattern LABEL = Pattern.compile("^BMX3_(I2R|R2I)_TRAFFIC_(\\d+)$");

  /**
   * @param args one path, to keylog.json
   * @throws Exception on I/O failure
   */
  public static void main(String[] args) throws Exception {
    if(args.length != 1) {
      System.err.println("usage: KeylogCheck <keylog.json>");
      System.exit(2);
    }
    JSONObject doc = new JSONObject(
        new String(Files.readAllBytes(Paths.get(args[0])), StandardCharsets.UTF_8));
    JSONArray capture = doc.optJSONArray("capture");
    JSONArray expected = doc.optJSONArray("expected");
    if(capture == null || expected == null || capture.length() == 0
        || capture.length() != expected.length()) {
      System.err.println("vector malformed");
      System.exit(1);
    }

    Map<String, byte[]> keys = parseKeylog(doc.optJSONArray("keylog"));
    if(keys.isEmpty()) {
      System.err.println("no usable key-log entries in the vector");
      System.exit(1);
    }

    int failures = 0;
    for(int i = 0; i < capture.length(); i++) {
      JSONObject frame = capture.getJSONObject(i);
      JSONObject want = expected.getJSONObject(i);
      String dir = frame.getString("dir");
      long seq = frame.getJSONObject("frame").getLong("seq");
      byte[] ct = Base64.getDecoder().decode(frame.getJSONObject("frame").getString("ct"));
      byte[] key = keys.get(dir + ":" + want.getInt("epoch"));
      if(key == null) {
        System.out.printf("FAIL frame %d: no key for %s epoch %d%n", i, dir, want.getInt("epoch"));
        failures++;
        continue;
      }
      byte[] pt;
      try {
        pt = TransportSession.openCiphertext(key, seq, ct);
      } catch(Exception e) {
        System.out.printf("FAIL frame %d: the logged %s key did not open it%n", i, dir);
        failures++;
        continue;
      }
      // JSONObject.similar compares structurally, so member order does not
      // matter -- the contract is the structure, not the text.
      JSONObject got = new JSONObject(new String(pt, StandardCharsets.UTF_8));
      if(got.similar(want.getJSONObject("inner"))) {
        System.out.printf("PASS frame %d (%s seq %d)%n", i, dir, seq);
      } else {
        System.out.printf("FAIL frame %d%n  got:  %s%n  want: %s%n",
            i, got, want.getJSONObject("inner"));
        failures++;
      }
    }

    // Self-test the oracle: a ciphertext no key seals must be refused, or a
    // checker that reported success for everything would look identical to this.
    byte[] anyKey = keys.values().iterator().next();
    boolean refused = false;
    try {
      TransportSession.openCiphertext(anyKey, 0L, new byte[32]);
    } catch(Exception e) {
      refused = true;
    }
    if(refused) System.out.println("PASS self-test: an unopenable frame is refused");
    else {
      System.out.println("FAIL self-test: an unopenable frame was accepted");
      failures++;
    }

    if(failures > 0) {
      System.err.println(failures + " key-log frame(s) failed");
      System.exit(1);
    }
    System.out.println("every captured frame opens with its logged key and reproduces the vector");
  }

  // '#' lines are comments; an unknown label shape is ignored rather than fatal,
  // so a future label is not a breaking change.
  private static Map<String, byte[]> parseKeylog(JSONArray lines) {
    Map<String, byte[]> keys = new HashMap<>();
    if(lines == null) return keys;
    for(int i = 0; i < lines.length(); i++) {
      String line = lines.getString(i).trim();
      if(line.isEmpty() || line.startsWith("#")) continue;
      String[] parts = line.split("\\s+");
      if(parts.length != 3) continue;
      Matcher m = LABEL.matcher(parts[0]);
      if(!m.matches()) continue;
      byte[] key;
      try {
        key = HEX.parseHex(parts[2]);
      } catch(IllegalArgumentException e) {
        continue;
      }
      if(key.length != 32) continue;
      keys.put(m.group(1).toLowerCase() + ":" + m.group(2), key);
    }
    return keys;
  }
}
